### ISSUE TITLE
wgengine/magicsock: seeing disco pings to IPv6-mapped-IPv4 addresses

### ISSUE THREAD

**Issue by bradfitz** [tailscale/tailscale#5602]
Noticed on Linux at HEAD (16939f0d56e4543a5f9e3),

```
ping(100.108.228.17): sending disco ping to [XvMEx] ub2004desk ...
magicsock: disco: failed to send *disco.Ping to [::ffff:209.180.x.y]:46078: write udp6 [::]:56726->[::ffff:209.180.x.y]:46078: sendto: network is unreachable
magicsock: disco: node [XvMEx] d:b508a2efe1de7f28 now using 10.2.200.239:46078
```

Note the ipv6-mapped ipv4 addresses there. I suspect something (maybe the new BPF disco stuff?) isn't unmapping somewhere.

/cc @raggi @danderson

**Comment by bradfitz** [comment]
Related to STUN somewhere:

> magicsock: endpoints changed: 209.180.x.y:50451 (stun), [::ffff:209.180.x.y]:50451 (stun), 10.0.0.52:50451 (local), 10.2.0.52:50451 (local), 172.17.0.1:50451 (local), [2602:xx:...:4148]:50451 (local)

**Comment by bradfitz** [comment]
It's from:

```
	if nr.GlobalV6 != "" {
		addAddr(ipp(nr.GlobalV6), tailcfg.EndpointSTUN)
	}
```

That `GlobalV6` string is literally a 4-in-6 ipport string:

```
panic: BOOM got 4 in 6 from "[::ffff:209.180.x.y]:48895"

goroutine 302 [running]:
tailscale.com/wgengine/magicsock.(*Conn).determineEndpoints.func1({0xc0008ba140, 0x1e})
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:1075 +0xc5
tailscale.com/wgengine/magicsock.(*Conn).determineEndpoints(0xc000158380, {0xf32bb8, 0xc0004aec40})
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:1116 +0x379
tailscale.com/wgengine/magicsock.(*Conn).updateEndpoints(0xc000158380, {0xdc8080, 0xe})
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:672 +0x1f0
created by tailscale.com/wgengine/magicsock.(*Conn).updateEndpoints.func1
        /home/bradfitz/src/tailscale.com/wgengine/magicsock/magicsock.go:631 +0x17b
```

**Comment by danderson** [comment]
STUN packets don't go through the BPF receive path, we filter on the disco magic number. So this sounds like we've not been unmapping netcheck data since switching to net/netip?

**Comment by bradfitz** [comment]
Ah, stacks suggests it's from 8f38afbf8e2ac0e6b516c99150bb5253012017de.

cc @skriptble

**Comment by bradfitz** [comment]
Ah, yes, it used to unmap but now it doesn't:

https://github.com/tailscale/tailscale/commit/8f38afbf8e2ac0e6b516c99150bb5253012017de#diff-f9c9570d6425603afd944c122b0805ca6d8612f079142adca496bee42051ee41L282

I guess the assumption was the stun package would do that, but it doesn't.

**Comment by skriptble** [comment]
Yeah, I just missed that we were unmapping here. I can put up a PR to fix this.

**Comment by bradfitz** [comment]
Sorry, I'd already prepared https://github.com/tailscale/tailscale/pull/5603 while debugging and didn't see GitHub.

I also checked control and unfortunately since 1.30.0 clients have been sending their endpoints to the control plane as these addresses which is kinda interfering with NAT traversal, at least adding some latency to it.

We should also unmap server side, in addition to cherry-picking this fix to 1.30.2.

/cc @DentonGentry

**Comment by bradfitz** [comment]
We've fixed this as much as we can on the control plane. The fix is cherry-picked for 1.30.2.
